cmd/openshift-install/graph.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/awalterschulze/gographviz"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/cluster"
+	"github.com/openshift/installer/pkg/asset/ignition/bootstrap"
+	"github.com/openshift/installer/pkg/asset/ignition/machine"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/manifests"
+	"github.com/openshift/installer/pkg/asset/metadata"
+)
+
+var (
+	graphOpts struct {
+		outputFile string
+	}
+)
+
+func newGraphCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "graph",
+		Short: "Outputs the internal dependency graph for installer",
+		Long:  "",
+		RunE:  runGraphCmd,
+	}
+	cmd.PersistentFlags().StringVar(&graphOpts.outputFile, "output-file", "", "file where the graph is written, if empty prints the graph to Stdout.")
+	return cmd
+}
+
+func runGraphCmd(cmd *cobra.Command, args []string) error {
+	g := gographviz.NewGraph()
+	g.SetName("G")
+	g.SetDir(true)
+	g.SetStrict(true)
+	targets := []struct {
+		name string
+		deps []asset.Asset
+	}{{
+		name: `"Install Config"`,
+		deps: []asset.Asset{&installconfig.InstallConfig{}},
+	}, {
+		name: `"Manifests"`,
+		deps: []asset.Asset{&manifests.Manifests{}, &manifests.Tectonic{}},
+	}, {
+		name: `"Ignition Configs"`,
+		deps: []asset.Asset{&bootstrap.Bootstrap{}, &machine.Master{}, &machine.Worker{}},
+	}, {
+		name: `"Cluster"`,
+		deps: []asset.Asset{&cluster.TerraformVariables{}, &kubeconfig.Admin{}, &cluster.Cluster{}, &metadata.Metadata{}},
```

Comment:
To stay DRY, we should put these lists under variables in `targets.go`.  Maybe:

```go
type target struct {
  name string
  command *cobra.Command
  assets []asset.Asset
}

var targets := map[string]target{
  "install-config": {
    name: "Install Config",
    command: &cobra.Command{
      Use:   "install-config",
      Short: "Generates the Install Config asset",
    },
    assets: {&installconfig.InstallConfig{}},
  },
  ...
}
```

Then in `main()`:

```go
for key, target := range targets { // maybe after sorting?
  target.command.RunE = runTargetCmd(target.assets...)
  subCmds = append(subCmds, target.command)
}
```

Reply:
done.